### PR TITLE
add result to after hook

### PR DESCRIPTION
**Pull request by Tomazh**
Hi, this is my first time making a PR here =)

I have an use case where I need to persist changes made on documents in a new collection and the after hook seemed like the perfect place to do that.
However this hook lacks the result from the `findOneAndUpdate` method, so I added it in there.
I also notice that there was not much going on with the variable `context` so I replaced it with the result

### CODE REVIEW

File: src/model.ts [hunk 1]
```diff
@@ -154,18 +154,17 @@ function wrap<A, R>(
   return async function modelWrapped(...args) {
     const { collectionName } = model.collection;
     const { after = [], before = [] } = model.options?.hooks || {};
-    const context = {};
```
Comment (avaly) [inline src/model.ts:157]: Please do not remove the `context` variable. We rely on it at Plex to run some custom logic.

If you want to achieve your intended changes with this PR, just attach `result` in the `context` object.

**Review by avaly — CHANGES_REQUESTED** [review]
This reminds me that we should document the hooks in detail.

**Comment by Tomazh** [comment]
Ok I added back the context variable.
I also changed the position of the `result` parameter and made it optional so it wont be a breaking change when it gets deployed.

**Review by avaly — COMMENTED** [review]
I still think that maybe the best way to implement this would be to attach `result` in the context object on line 166:

```ts
context.result = result;
```

File: src/hooks.ts [hunk 2]
```diff
@@ -7,7 +7,7 @@ export type Hook<A> = (
   methodName: string,
   args: A[],
   context: unknown,
-  error?: Error
+  error?: Error,
```
Comment (avaly) [inline src/hooks.ts:10]: nit: please revert this change

File: src/__tests__/model.test.ts [hunk 3]
```diff
@@ -60,7 +60,7 @@ describe('model', () => {
 
   beforeEach(() => {
     doc = {
-      _id: new ObjectId(),
+      _id: new ObjectId('62b3876eb20d64c7bb577874'),
```
Comment (avaly) [inline src/__tests__/model.test.ts:63]: Why is this ID needed here?

**Comment by marafat** [comment]
I don't object to the idea but I feel `result` should be a first class argument on the hook just like `error`. This might mean that we will need a separate type for `HookAfter` and `HookBefore`, and it is going to be a breaking change, but I feel this is the right way to do it. 

`context` is not meant to have any typed members of the library. It is just a mean for the consumer of the library to pass some values in the before hooks and act on it in the after hook.

**Comment by Tomazh** [comment]
Yeah I agree that it makes more sense to have a separate `result` type on each hook. We could avoid adding a breacking change by adding a overload of the hook function, what do you think?

**Comment by Tomazh** [comment]
Yeah I agree that it makes more sense to add a separate type for hooks. We could also add `result` as an optional parameter right after `error`. Not pretty, I know, but it would avoid adding breaking changes:

src/hooks.ts
```
export type Hook<A> = (
  collectionName: string,
  methodName: string,
  args: A[],
  context: unknown,
  error?: Error,
  result?: unknown
) => Promise<void>;
```
src/model.ts
`await hook(collectionName, method.name, args, context, undefined, result );`

File: src/__tests__/model.test.ts (hunk 3 above)
Comment (Tomazh) [reply src/__tests__/model.test.ts:63]: You're right, not needed anymore, I wasn't using the `result` variable to test the input parameters at first and was hard coding the expected input.

**Comment by Tomazh** [comment]
Awesome, I'll wait for @avaly to give his opinion on this approach before making any changes

**Comment by avaly** [comment]
I see your arguments for keeping the `result` separate from the `context` @marafat.

If we do add them as a separate argument to those hooks functions, and we're going to make this a breaking change anyway, let's transform the positional arguments for the hooks functions into a single object argument. This way any future additions to the input of the hooks, won't trigger a major version bump.

```ts
export type Hook<A> = (params: {
  args: A[];
  collectionName: string;
  context: unknown;
  error?: Error;
  methodName: string;
  result?: unknown;
}) => Promise<void>;
```

**Review by avaly — DISMISSED** [review]
This looks great! Thanks for the contribution!

File: src/hooks.ts [hunk 4]
```diff
@@ -3,11 +3,14 @@ import { inspect } from 'util';
...
-  args: A[],
-  context: unknown,
-  error?: Error
+  params: {
+    args: A[];
+    collectionName: string;
+    context: unknown;
+    error?: Error;
+    methodName: string;
+    result?: unknown;
```
Comment (marafat) [inline src/hooks.ts:12]: I wonder if we can type this from the model 🤔

**Review by marafat — COMMENTED** [review]
@avaly spends more time here than  I do, so if it looks good for him I am ok with it. I did notice some lint issues on unchanged files, worth taking a look at it. 
CleanShot 2022-06-28 at 11 07 28


Regarding my note on typing the `result`, I don't think it is a blocker but it is a nice follow up.

**Comment by avaly** [comment]
> I did notice some lint issues on unchanged files, worth taking a look at it.

Nevermind those. They are old issues. We should suppress them at some point.

**Review by avaly — APPROVED** [review]

**Comment by Tomazh** [comment]
Tks for all the help @avaly and @marafat!
I don't have write access to merge this PR, could one of you do the merge for me?
